sql: fix UNIQUE column definition for REGIONAL BY ROW

otan opened the pull request:
This was buggy as the implicitly added crdb_region column was not
initialized before the UNIQUE definition was available. Instead, defer
these index creations until after all column definitions have been
initialized.

Release justification: low-risk bug fix to new functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/61449)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-:-MUxbpD35viWKwu0YbDU:b-g2a5fb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm and @otan)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 1799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxaMGj3Mcbmd9p1BKc:-MUxaMGj3Mcbmd9p1BKd:b-3zzld4) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L1799)):*
> ```
> 
> statement ok
> CREATE TABLE regional_by_row_unique_in_column_def (
> ```

[nit] I think you want to name this table ....abc instead of ....def
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 1823 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxaigL8GBzeXJ_OoLO:-MUxaigL8GBzeXJ_OoLP:b-a2jgdv) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L1823)):*
> ```
> CREATE TABLE regional_by_row_fk_reference (
>   d INT PRIMARY KEY,
>   e INT REFERENCES regional_by_row_unique_in_column_def(a),
> ```

How about also adding UNIQUE for one of these here? e.g.,
`e INT UNIQUE REFERENCES regional_by_row_unique_in_column_def(a),`
___
*[pkg/sql/create_table.go, line 1809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxbInS42fZJPh23VhP:-MUxbInS42fZJPh23VhQ:b2pj25x) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/sql/create_table.go#L1809)):*
> ```Go
> 
> 	for _, implicitColumnDefIdx := range implicitColumnDefIdxs {
> 		// If it a non-primary index that is implicitly created, ensure partitioning
> ```

[nit] If it a -> If it is a


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61449)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajstorm and @rytaft)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 1799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxaMGj3Mcbmd9p1BKc:-MUyfz_v-4Ei41IwN8qf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L1799)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] I think you want to name this table ....abc instead of ....def
</blockquote></details>

Done.
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 1823 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxaigL8GBzeXJ_OoLO:-MUyg-3P1V1l0C7Kp_BZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L1823)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

How about also adding UNIQUE for one of these here? e.g.,
`e INT UNIQUE REFERENCES regional_by_row_unique_in_column_def(a),`
</blockquote></details>

Done.
___
*[pkg/sql/create_table.go, line 1809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxbInS42fZJPh23VhP:-MUyg-b14sjMX_fY8gGJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/sql/create_table.go#L1809)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] If it a -> If it is a
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
sql: fix UNIQUE column definition for REGIONAL BY ROW

This was buggy as the implicitly added crdb_region column was not
initialized before the UNIQUE definition was available. Instead, defer
these index creations until after all column definitions have been
initialized.

Release justification: low-risk bug fix to new functionality

Release note: None

rytaft reviewed (approved):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/61449)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajstorm)

___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 1799 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxaMGj3Mcbmd9p1BKc:-MUz01LzDPvPajtlyDlL:b-6qpauc) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L1799)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Oh sorry -- I thought you were naming it after the columns and was confused about why you said "def" instead of "abc". Feel free to change it back to _def if you prefer (I see now that you were referring to the ColumnDef from the AST). Fine with me either way
___
*[pkg/ccl/logictestccl/testdata/logic_test/regional_by_row, line 1823 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/61449#-MUxaigL8GBzeXJ_OoLO:-MUz0f323K3iGdrkLdJl:b9rujn2) ([raw file](https://github.com/cockroachdb/cockroach/blob/284305e85f2a4ec3cad005f0c54c58195a767d5b/pkg/ccl/logictestccl/testdata/logic_test/regional_by_row#L1823)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

No need to make them both unique -- might be nice to have one of each (but again -- I don't have a strong opinion here)


<!-- Sent from Reviewable.io -->

otan commented:
hah it's fine, i'll live without the commit extra round trip :D 

bors r=rytaft
